# Analyzing cars selling in Ebay Platform
### The main goal of the project is to analyze the included used car listings.
### In order to do that we have downloaded a dataset with 50,000 data points from a section of the German eBay website.
#### We firstly will clean the data and then analyze the included used car listings.

In [871]:
import numpy as np
import pandas as pd
autos = pd.read_csv('D:/PROYECTOS_PYTHON/autos.csv', encoding = 'Latin-1')

Once opened the file, let's have a look in it!

In [872]:
autos.head()
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

As we can see, there are 20 columns (most of them are strings) and 371528 rows. Some of the columns have some null values (index 6,8,10,13 and 15) but none have more than 20% null values.
As the column names use camelcase instead of Python's preferred snakecase we are going to change it.

In [873]:
column_names = autos.columns
print(column_names)
dict_columns = {'dateCrawled': 'date_crawled',
                'name':'name',
                'seller':'seller',
                'offerType':'offer_type',
                'price':'price',
                'abtest':'abtest',
                'vehicleType':'vehicle_type',
                'yearOfRegistration':'registration_year',
                'gearbox':'gearbox',
                'powerPS':'power_ps',
                'model':'model',
                'kilometer':'kilometer',
                'monthOfRegistration':'registration_month',
                'fuelType':'fuel_type',
                'brand':'brand',
                'notRepairedDamage':'unrepaired_damage',
                'dateCreated':'ad_created',
                'nrOfPictures':'nr_pictures',
                'postalCode':'postal_code',
                'lastSeen':'last_seen'}
autos.columns = column_names.map(dict_columns)
print(autos.columns)


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')
Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [874]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00            0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00            0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00            0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00            0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00            0        60437  2016-04-06 10:17:21

## Exploring and cleaning the data

Now, we are going to explore deeper the data:

In [875]:
autos.describe(include='all')

date_crawled         name  seller offer_type  \
count                371528       371528  371528     371528   
unique               280500       233531       2          2   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot   
freq                      7          657  371525     371516   
mean                    NaN          NaN     NaN        NaN   
std                     NaN          NaN     NaN        NaN   
min                     NaN          NaN     NaN        NaN   
25%                     NaN          NaN     NaN        NaN   
50%                     NaN          NaN     NaN        NaN   
75%                     NaN          NaN     NaN        NaN   
max                     NaN          NaN     NaN        NaN   

                     price  abtest vehicle_type  registration_year  gearbox  \
count        371,528.00000  371528       333659      371,528.00000   351319   
unique                 nan       2            8                nan        2   
top                    nan    test    limousine                nan  manuell   
freq                   nan  192585        95894                nan   274214   
mean          17,295.14187     NaN          NaN        2,004.57800      NaN   
std        3,587,953.74441     NaN          NaN           92.86660      NaN   
min                0.00000     NaN          NaN        1,000.00000      NaN   
25%            1,150.00000     NaN          NaN        1,999.00000      NaN   
50%            2,950.00000     NaN          NaN        2,003.00000      NaN   
75%            7,200.00000     NaN          NaN        2,008.00000      NaN   
max    2,147,483,647.00000     NaN          NaN        9,999.00000      NaN   

            power_ps   model     kilometer  registration_month fuel_type  \
count  371,528.00000  351044 371,528.00000       371,528.00000    338142   
unique           nan     251           nan                 nan         7   
top              nan    golf           nan                 nan    benzin   
freq             nan   30070           nan                 nan    223857   
mean       115.54948     NaN 125,618.68823             5.73445       NaN   
std        192.13958     NaN  40,112.33705             3.71241       NaN   
min          0.00000     NaN   5,000.00000             0.00000       NaN   
25%         70.00000     NaN 125,000.00000             3.00000       NaN   
50%        105.00000     NaN 150,000.00000             6.00000       NaN   
75%        150.00000     NaN 150,000.00000             9.00000       NaN   
max     20,000.00000     NaN 150,000.00000            12.00000       NaN   

             brand unrepaired_damage           ad_created   nr_pictures  \
count       371528            299468               371528 371,528.00000   
unique          40                 2                  114           nan   
top     volkswagen              nein  2016-04-03 00:00:00           nan   
freq         79640            263182                14450           nan   
mean           NaN               NaN                  NaN       0.00000   
std            NaN               NaN                  NaN       0.00000   
min            NaN               NaN                  NaN       0.00000   
25%            NaN               NaN                  NaN       0.00000   
50%            NaN               NaN                  NaN       0.00000   
75%            NaN               NaN                  NaN       0.00000   
max            NaN               NaN                  NaN       0.00000   

         postal_code            last_seen  
count  371,528.00000               371528  
unique           nan               182806  
top              nan  2016-04-06 13:45:54  
freq             nan                   17  
mean    50,820.66764                  NaN  
std     25,799.08247                  NaN  
min      1,067.00000                  NaN  
25%     30,459.00000                  NaN  
50%     49,610.00000                  NaN  
75%     71,546.00000                  NaN  
max     99,

In [876]:
autos["seller"].value_counts()

privat        371525
gewerblich         3
Name: seller, dtype: int64

In [877]:
autos["offer_type"].value_counts()

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64

In [878]:
autos["abtest"].value_counts()

test       192585
control    178943
Name: abtest, dtype: int64

In [879]:
autos["gearbox"].value_counts()

manuell      274214
automatik     77105
Name: gearbox, dtype: int64

In [880]:
autos["unrepaired_damage"].value_counts()

nein    263182
ja       36286
Name: unrepaired_damage, dtype: int64

In [881]:
autos["nr_pictures"].value_counts()

0    371528
Name: nr_pictures, dtype: int64

As we can observe, the columns seller and offer_type have mostly one unique value, so they are candidates to be dropped.
The column nr_pictures has only one value, therefore we will drop it later.
We are going to change the postal_code type because although it's a number, we are not going to make any operations and a number bigger than other doesn't mean that has more value:

In [882]:
autos["postal_code"] = autos["postal_code"].astype(str)
autos.describe(include='all')

date_crawled         name  seller offer_type  \
count                371528       371528  371528     371528   
unique               280500       233531       2          2   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot   
freq                      7          657  371525     371516   
mean                    NaN          NaN     NaN        NaN   
std                     NaN          NaN     NaN        NaN   
min                     NaN          NaN     NaN        NaN   
25%                     NaN          NaN     NaN        NaN   
50%                     NaN          NaN     NaN        NaN   
75%                     NaN          NaN     NaN        NaN   
max                     NaN          NaN     NaN        NaN   

                     price  abtest vehicle_type  registration_year  gearbox  \
count        371,528.00000  371528       333659      371,528.00000   351319   
unique                 nan       2            8                nan        2   
top                    nan    test    limousine                nan  manuell   
freq                   nan  192585        95894                nan   274214   
mean          17,295.14187     NaN          NaN        2,004.57800      NaN   
std        3,587,953.74441     NaN          NaN           92.86660      NaN   
min                0.00000     NaN          NaN        1,000.00000      NaN   
25%            1,150.00000     NaN          NaN        1,999.00000      NaN   
50%            2,950.00000     NaN          NaN        2,003.00000      NaN   
75%            7,200.00000     NaN          NaN        2,008.00000      NaN   
max    2,147,483,647.00000     NaN          NaN        9,999.00000      NaN   

            power_ps   model     kilometer  registration_month fuel_type  \
count  371,528.00000  351044 371,528.00000       371,528.00000    338142   
unique           nan     251           nan                 nan         7   
top              nan    golf           nan                 nan    benzin   
freq             nan   30070           nan                 nan    223857   
mean       115.54948     NaN 125,618.68823             5.73445       NaN   
std        192.13958     NaN  40,112.33705             3.71241       NaN   
min          0.00000     NaN   5,000.00000             0.00000       NaN   
25%         70.00000     NaN 125,000.00000             3.00000       NaN   
50%        105.00000     NaN 150,000.00000             6.00000       NaN   
75%        150.00000     NaN 150,000.00000             9.00000       NaN   
max     20,000.00000     NaN 150,000.00000            12.00000       NaN   

             brand unrepaired_damage           ad_created   nr_pictures  \
count       371528            299468               371528 371,528.00000   
unique          40                 2                  114           nan   
top     volkswagen              nein  2016-04-03 00:00:00           nan   
freq         79640            263182                14450           nan   
mean           NaN               NaN                  NaN       0.00000   
std            NaN               NaN                  NaN       0.00000   
min            NaN               NaN                  NaN       0.00000   
25%            NaN               NaN                  NaN       0.00000   
50%            NaN               NaN                  NaN       0.00000   
75%            NaN               NaN                  NaN       0.00000   
max            NaN               NaN                  NaN       0.00000   

       postal_code            last_seen  
count       371528               371528  
unique        8150               182806  
top          10115  2016-04-06 13:45:54  
freq           828                   17  
mean           NaN                  NaN  
std            NaN                  NaN  
min            NaN                  NaN  
25%            NaN                  NaN  
50%            NaN                  NaN  
75%            NaN                  NaN  
max            NaN               

In [883]:
autos.shape

(371528, 20)

Since a one value column doesn't add valuable information we are going to drop "nr_pictures":

In [884]:
autos.drop(columns=["nr_pictures"], inplace=True)

In [885]:
autos.describe(include='all')


date_crawled         name  seller offer_type  \
count                371528       371528  371528     371528   
unique               280500       233531       2          2   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot   
freq                      7          657  371525     371516   
mean                    NaN          NaN     NaN        NaN   
std                     NaN          NaN     NaN        NaN   
min                     NaN          NaN     NaN        NaN   
25%                     NaN          NaN     NaN        NaN   
50%                     NaN          NaN     NaN        NaN   
75%                     NaN          NaN     NaN        NaN   
max                     NaN          NaN     NaN        NaN   

                     price  abtest vehicle_type  registration_year  gearbox  \
count        371,528.00000  371528       333659      371,528.00000   351319   
unique                 nan       2            8                nan        2   
top                    nan    test    limousine                nan  manuell   
freq                   nan  192585        95894                nan   274214   
mean          17,295.14187     NaN          NaN        2,004.57800      NaN   
std        3,587,953.74441     NaN          NaN           92.86660      NaN   
min                0.00000     NaN          NaN        1,000.00000      NaN   
25%            1,150.00000     NaN          NaN        1,999.00000      NaN   
50%            2,950.00000     NaN          NaN        2,003.00000      NaN   
75%            7,200.00000     NaN          NaN        2,008.00000      NaN   
max    2,147,483,647.00000     NaN          NaN        9,999.00000      NaN   

            power_ps   model     kilometer  registration_month fuel_type  \
count  371,528.00000  351044 371,528.00000       371,528.00000    338142   
unique           nan     251           nan                 nan         7   
top              nan    golf           nan                 nan    benzin   
freq             nan   30070           nan                 nan    223857   
mean       115.54948     NaN 125,618.68823             5.73445       NaN   
std        192.13958     NaN  40,112.33705             3.71241       NaN   
min          0.00000     NaN   5,000.00000             0.00000       NaN   
25%         70.00000     NaN 125,000.00000             3.00000       NaN   
50%        105.00000     NaN 150,000.00000             6.00000       NaN   
75%        150.00000     NaN 150,000.00000             9.00000       NaN   
max     20,000.00000     NaN 150,000.00000            12.00000       NaN   

             brand unrepaired_damage           ad_created postal_code  \
count       371528            299468               371528      371528   
unique          40                 2                  114        8150   
top     volkswagen              nein  2016-04-03 00:00:00       10115   
freq         79640            263182                14450         828   
mean           NaN               NaN                  NaN         NaN   
std            NaN               NaN                  NaN         NaN   
min            NaN               NaN                  NaN         NaN   
25%            NaN               NaN                  NaN         NaN   
50%            NaN               NaN                  NaN         NaN   
75%            NaN               NaN                  NaN         NaN   
max            NaN               NaN                  NaN         NaN   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Now, we are going to explore the strange values in "seller" and "offer_type" columns:

In [886]:
seller_gewerblich = autos[autos["seller"] == "gewerblich"]
seller_gewerblich

date_crawled                                         name  \
59155   2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   
149393  2016-03-22 09:54:43              Chevrolet_Matiz_1.Hd_TÜV11/2017   
186181  2016-03-14 15:58:22               BMW_2000_AUTOMATIK_H_ZULASSUNG   

            seller offer_type  price   abtest vehicle_type  registration_year  \
59155   gewerblich    Angebot    100  control        kombi               2000   
149393  gewerblich    Angebot   1100     test   kleinwagen               2006   
186181  gewerblich    Angebot   6900  control    limousine               1967   

          gearbox  power_ps   model  kilometer  registration_month fuel_type  \
59155     manuell         0  megane     150000                   8    benzin   
149393    manuell        38   matiz     150000                  10    benzin   
186181  automatik       101  andere      70000                   8    benzin   

            brand unrepaired_damage           ad_created postal_code  \
59155     renault               NaN  2016-03-15 00:00:00       65232   
149393  chevrolet              nein  2016-03-22 00:00:00        7973   
186181        bmw              nein  2016-03-14 00:00:00       73614   

                  last_seen  
59155   2016-04-06 17:15:37  
149393  2016-04-05 21:15:48  
186181  2016-04-05 19:45:06

In [887]:
offertype_Gesuch= autos[autos["offer_type"] == "Gesuch"]
offertype_Gesuch

date_crawled  \
16744   2016-04-03 15:48:33   
90921   2016-03-27 14:55:45   
144154  2016-03-14 18:51:39   
150400  2016-03-07 21:37:03   
164788  2016-03-28 15:56:38   
183716  2016-03-11 10:57:56   
251359  2016-03-21 07:55:59   
285243  2016-03-15 18:55:15   
315232  2016-03-28 19:55:44   
327204  2016-03-13 07:52:29   
357794  2016-03-25 20:50:32   
360583  2016-03-26 14:42:50   

                                                     name  seller offer_type  \
16744                                Suche_VW_T5_Multivan  privat     Gesuch   
90921   Suche_einen_originalen_VW_Golf_2_Fire_and_Ice_G60  privat     Gesuch   
144154  [SUCHE]_BMW_e46_Coupe_330ci_M_Paket_Facelift__...  privat     Gesuch   
150400  Suche_Mercedes_300D__W124__Silberdistel_metall...  privat     Gesuch   
164788                                  Suche_bitte_hilfe  privat     Gesuch   
183716   Clio_3_Bremsen_vorne_und_hinten_wechseln__EILT!!  privat     Gesuch   
251359           SZUKAM_taniego_TRANSPORTU_auta_do_Polski  privat     Gesuch   
285243                                Trabant_gesucht!!!!  privat     Gesuch   
315232                 Ford_Fiesta_2005_mit_Unfallschaden  privat     Gesuch   
327204                  Kaufe_Bastlerfahrzeuge_aller_art.  privat     Gesuch   
357794          Suche_BMW_e46_M3_Coupé_!_kein_M1_M4_M5_M6  privat     Gesuch   
360583  Tausche_oder_verkaufe_Ford_Galaxy_2_3_Liter_Be...  privat     Gesuch   

        price   abtest vehicle_type  registration_year  gearbox  power_ps  \
16744       0     test          bus               2005      NaN         0   
90921   19999  control          NaN               1990      NaN         0   
144154      0     test        coupe               2004  manuell         0   
150400      0  control          NaN               1985      NaN       109   
164788      0  control          NaN               2016  manuell         0   
183716      0     test   kleinwagen               2005  manuell         0   
251359    150  control          NaN               2000      NaN         0   
285243   2000  control          NaN               1980      NaN         0   
315232   1550  control   kleinwagen               2005  manuell        90   
327204    555     test          NaN               2000  manuell        55   
357794  20000  control          NaN               2005      NaN         0   
360583      0  control          bus               1998  manuell       147   

              model  kilometer  registration_month fuel_type           brand  \
16744   transporter     150000                   0       NaN      volkswagen   
90921          golf       5000                   0       NaN      volkswagen   
144154          3er     150000                   1    benzin             bmw   
150400     e_klasse     150000                   0       NaN   mercedes_benz   
164788        astra      20000                   2    diesel            opel   
183716         clio     150000                  12    benzin         renault   
251359          NaN       5000                   1       NaN  sonstige_autos   
285243       andere      50000                   0       NaN         trabant   
315232       fiesta     150000                   3    diesel            ford   
327204          NaN     150000                   3       NaN  sonstige_autos   
357794          NaN       5000                   0       NaN             bmw   
360583       galaxy     150000                   6    benzin            ford   

       unrepaired_damage           ad_created postal_code            last_seen  
16744                NaN  2016-04-03 00:00:00       29690  2016-04-05 15:16:06  
90921                NaN  2016-03-27 00:00:00       50181  2016-04-07 11:45:18  
144154              nein  2016-03-14 00:00:00       86956  2016-03-25 06:17:06  
150400               NaN  2016-03-07 00:00:00       68169  2016-03-07 21:37:03  
164788              nein  2016-03-28 00:00:00       26215  2016-03-28 15:56:38  
183716              nein  2016-03-11 00:0

We are going to focus in private sellers because there are only 3 rows with this kind of sellers.
Same as the observation for seller, we are going to exclude the request (Gesuch) data of the offer_type column in order to avoid that it messes up our analysis.\
For that, we create firstly a function to get the index numbers of this rows in order to drop them later.

In [888]:
def get_index_positions_2(list_of_elems, element):
    ''' Returns the indexes of all occurrences of give element in
    the list- listOfElements '''
    index_pos_list = []
    for i in range(len(list_of_elems)):
        if list_of_elems[i] == element:
            index_pos_list.append(i)
    return index_pos_list

In [889]:
index_Gesuch = get_index_positions_2(autos["offer_type"], 'Gesuch')
index_Gesuch

[16744,
 90921,
 144154,
 150400,
 164788,
 183716,
 251359,
 285243,
 315232,
 327204,
 357794,
 360583]

In [890]:
index_gewerblich = get_index_positions_2(autos["seller"], 'gewerblich')
index_gewerblich

[59155, 149393, 186181]

In [891]:
autos.drop(index=index_Gesuch, inplace=True)

In [892]:
autos.drop(index=index_gewerblich, inplace=True)

In [893]:
autos.shape

(371513, 19)

As we can see, we have 1 column and 15 rows less.

## Exploring prices and kilometers rows

Next, we are going to explore price and kilometer columns to detect outliers.\
In order to read the data easily we change the display format:

In [894]:
pd.options.display.float_format = '{:,.2f}'.format

In [895]:
print(autos["price"].describe())
print('\n')
print(autos["price"].shape)
print('\n')
print(autos["price"].value_counts())
print('\n')
print('The series price has ', autos["price"].unique().shape[0],' unique values ')

count         371,513.00
mean           17,295.70
std         3,588,026.18
min                 0.00
25%             1,150.00
50%             2,950.00
75%             7,200.00
max     2,147,483,647.00
Name: price, dtype: float64


(371513,)


0         10772
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price, Length: 5597, dtype: int64


The series price has  5597  unique values 


We observe in prices that the maximum value is incredibly high and it's obvious that is not real.\
We are going to explore the values higher than 1M:

In [896]:
values_more_2M = autos[autos["price"] > 2000000]
values_more_1Mto2M = autos[autos["price"].between(1000000, 2000000)]
print(values_more_2M.shape)

(56, 19)


In [897]:
values_more_2M

date_crawled  \
14663   2016-04-03 06:03:56   
20746   2016-03-08 15:54:52   
21467   2016-03-19 19:45:02   
39625   2016-03-08 23:53:51   
41980   2016-03-11 09:37:17   
45055   2016-04-02 00:06:21   
54295   2016-03-08 11:51:40   
54507   2016-03-09 11:40:15   
56973   2016-03-18 18:52:27   
69747   2016-03-20 10:59:32   
77520   2016-03-22 14:58:27   
87799   2016-03-08 20:39:05   
90853   2016-04-01 05:36:16   
91106   2016-03-30 19:46:05   
108987  2016-03-29 00:38:37   
120919  2016-03-25 12:43:02   
121793  2016-03-05 20:59:54   
127769  2016-03-28 12:51:51   
127820  2016-03-14 17:49:54   
129969  2016-03-11 08:37:57   
139052  2016-03-14 12:49:48   
150909  2016-04-02 22:41:54   
160530  2016-03-29 12:58:27   
171824  2016-03-29 23:52:57   
176070  2016-04-02 14:43:10   
177744  2016-03-31 16:37:00   
183361  2016-03-26 11:51:20   
185227  2016-03-14 16:52:51   
189737  2016-04-03 12:41:33   
196306  2016-04-01 01:59:20   
210657  2016-03-12 21:50:57   
217547  2016-03-11 10:40:52   
241847  2016-03-31 18:56:54   
246635  2016-03-25 11:55:36   
251980  2016-03-21 22:25:19   
266124  2016-04-03 23:36:16   
268173  2016-04-01 09:37:19   
273145  2016-03-09 15:45:47   
279937  2016-04-03 11:56:00   
280059  2016-03-05 20:49:28   
281833  2016-03-19 08:50:50   
287835  2016-03-26 00:56:36   
299872  2016-03-14 10:06:23   
310924  2016-03-14 05:36:22   
318143  2016-03-21 13:57:51   
325589  2016-03-27 21:55:14   
332028  2016-03-31 18:39:53   
338415  2016-03-26 22:49:39   
347435  2016-03-30 15:50:46   
358868  2016-03-16 16:49:53   
361451  2016-04-03 14:36:47   
362782  2016-03-06 10:38:42   
363434  2016-03-23 06:36:20   
364171  2016-04-04 21:25:21   
366653  2016-03-20 10:59:52   
366861  2016-03-21 10:53:06   

                                                     name  seller offer_type  \
14663                          Tausch_moeglich_Polo_vw_n9  privat    Angebot   
20746                                  Golf3_Schlachtfest  privat    Angebot   
21467   Porsche_911_911R_1_of_20_ORIGINAL_R_nur_798KG!...  privat    Angebot   
39625                         Tausche_volvo_v40_gegen_van  privat    Angebot   
41980                                       Maserati_MC12  privat    Angebot   
45055   Ferrari_LaFerrari_Deutsche_Auslieferung._A_Cel...  privat    Angebot   
54295                        Mazda_5_2.0_CD_DPF_Exclusive  privat    Angebot   
54507                    Tausch_DKW_F8_gegen_Lanz_Bulldog  privat    Angebot   
56973     tausche_ford_mondeo_mk3_ghia_gegen_anderes_auto  privat    Angebot   
69747                    Suche_Mercedes_Benz_r_klasse_cdi  privat    Angebot   
77520                         Tausch_gegen_gleichwertiges  privat    Angebot   
87799                                   Leasinguebernahme  privat    Angebot   
90853                               Verkaufe_oder_Tausche  privat    Angebot   
91106   Verkaufe_oder_tausche_wenn_nur_gegen_Kombi_s4_...  privat    Angebot   
108987  Biete_Hier_zum_Tauschen_an_aber_nur_realistisc...  privat    Angebot   
120919                ???_?_???____???_?_???____???_?_???  privat    Angebot   
121793                       Alter_VW_Kaefer_Scheunenfund  privat    Angebot   
127769                   Golf_4_laesst_sich_nicht_oeffnen  privat    Angebot   
127820                                   audi_a6_c5_avant  privat    Angebot   
129969                            Kaufe_AUTOS_Zum_basteln  privat    Angebot   
139052                Mercedes_Benz_CLS_350_CDI_7G_TRONIC  privat    Angebot   
150909                     Suche_Autos_Model_egal_welches  privat    Angebot   
160530                                     Golf_4_schrott  privat    Angebot   
171824  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  privat    Angebot   
176070                                      Suche_BMW_Fun  privat    Angebot   
177744                     Suche_Trabant_auf_Ratenzahlung  privat    Angebot   
183361  BMW_e60_nur_Tausch_prins_gasanlage_20_Zoll!_No...  privat    Angebot   
185227 

In [898]:
values_more_1Mto2M

date_crawled                                        name  \
16889   2016-03-15 16:45:45            Ford_Mondeo_Kombi_TÜV_abgelaufen   
20143   2016-03-18 00:55:53                  Porsche_911R_Einer_von_911   
45833   2016-03-15 10:53:11                     Porsche_911_Turbo__930_   
53507   2016-03-11 18:57:57  Suzuki_SJ_Samurai_413_Van__Kein_Lada_Niva_   
57594   2016-03-25 09:56:28                               Astra_f_teile   
78480   2016-03-10 20:43:12                               Porsche_991_R   
106651  2016-03-11 22:54:08               SUCHE_PKW_MIT_TÜV_MIN2_MONATE   
167750  2016-03-22 12:54:19                  Bmw_530d_zum_ausschlachten   
173279  2016-03-07 21:36:25                                 12345678911   
306469  2016-03-17 12:06:35                 Vw_Polo_86c_Schrott_Karosse   
312109  2016-04-05 11:50:18         A3_zu_verkaufen_fuer_1.000.000_Euro   
316200  2016-04-04 11:53:31                                 Ferrari_F40   
333699  2016-03-12 03:36:18                               Maserati_MC12   
367041  2016-03-24 19:00:03                               Maserati_MC12   

        seller offer_type    price   abtest vehicle_type  registration_year  \
16889   privat    Angebot  1000000  control        kombi               1998   
20143   privat    Angebot  1250000     test        coupe               2016   
45833   privat    Angebot  1000000  control        coupe               1981   
53507   privat    Angebot  1000000     test          suv               1992   
57594   privat    Angebot  1111111     test          NaN               1995   
78480   privat    Angebot  1250000     test        coupe               2016   
106651  privat    Angebot  1111111  control          NaN               2000   
167750  privat    Angebot  1234566  control        kombi               1999   
173279  privat    Angebot  1010010  control          NaN               2017   
306469  privat    Angebot  1000000  control        coupe               1985   
312109  privat    Angebot  1000000     test          NaN               2000   
316200  privat    Angebot  1300000  control        coupe               1992   
333699  privat    Angebot  1600000     test        coupe               2005   
367041  privat    Angebot  1600000     test        coupe               2005   

          gearbox  power_ps   model  kilometer  registration_month fuel_type  \
16889         NaN         0  mondeo     150000                   0    benzin   
20143     manuell       500     911       5000                   3    benzin   
45833     manuell       300     911     150000                   1    benzin   
53507     manuell        69  andere     100000                   0    benzin   
57594     manuell        71   astra     150000                   5    benzin   
78480     manuell       500     911       5000                   3    benzin   
106651        NaN         0     NaN     150000                   0       NaN   
167750  automatik       190     NaN     150000                   2    diesel   
173279    manuell       101  zafira     150000                   9    diesel   
306469    manuell        75    polo     150000                   0    benzin   
312109        NaN         0      a3     150000                   0       NaN   
316200        NaN         0     NaN      50000                  12       NaN   
333699        NaN         0     NaN      10000                   4       NaN   
367041        NaN       625     NaN      10000                   4    benzin   

                 brand unrepaired_damage           ad_created postal_code  \
16889             ford                ja  2016-03-15 00:00:00       26607   
20143          porsche              nein  2016-03-17 00:00:00       70435   
45833          porsche              nein  2016-03-15 00:00:00       10969   
53507           suzuki              nein  2016-03-11 00:00:00       45711   
57594             opel               NaN  2016-03-25 00:00:00       67700   
78480          porsche              nein

In both cases there are a lot of data probably false, and although we will lose real data of luxurious cars it's worthy to drop all the rows with a price higher than 1M or superior because they might disproportionately affect our observations.\
As we also can see, there are 10772 cars with a price of 0$ that probably is a fake information. Anyway, it's not useful information for our analysis therefore we will drop them too. 

In [899]:
autos = autos.drop(autos[(autos["price"]>=1000000)].index)

Now, let's explore data again:

In [900]:
print(autos["price"].describe())
print('\n')
print(autos["price"].shape)
print('\n')
print(autos["price"].value_counts())
print('\n')
print('The series price has ', autos["price"].unique().shape[0],' unique values ')

count   371,443.00
mean      5,811.84
std      11,946.76
min           0.00
25%       1,150.00
50%       2,950.00
75%       7,200.00
max     999,999.00
Name: price, dtype: float64


(371443,)


0        10772
500       5670
1500      5394
1000      4649
1200      4594
         ...  
33944        1
11149        1
17280        1
53650        1
32399        1
Name: price, Length: 5574, dtype: int64


The series price has  5574  unique values 


We can observe that we have lost less than 100 rows and 23 unique values therefore it is not a big loss and we mostly discard false information

We do the same with *kilometers* column:

In [901]:
print(autos["kilometer"].describe())
print('\n')
print(autos["kilometer"].shape)
print('\n')
print(autos["kilometer"].value_counts())
print('\n')
print('The series kilometer has ', autos["kilometer"].unique().shape[0],' unique values ')

count   371,443.00
mean    125,623.78
std      40,104.70
min       5,000.00
25%     125,000.00
50%     150,000.00
75%     150,000.00
max     150,000.00
Name: kilometer, dtype: float64


(371443,)


150000    240747
125000     38064
100000     15917
90000      12523
80000      11052
70000       9772
60000       8668
50000       7611
5000        7053
40000       6375
30000       6041
20000       5673
10000       1947
Name: kilometer, dtype: int64


The series kilometer has  13  unique values 


We observe that there are not outliers and the information could be true (the minimum and maximum values appear reasonable).\
We only have to take into account that it's not exact information, it's categorized information and the maximum value is 150000 so we probably don't have access to a detailed information about this.

## Exploring the date columns

In [902]:
autos["date_crawled"].value_counts(dropna=False)

2016-03-24 14:49:47    7
2016-03-19 21:49:56    6
2016-03-26 22:57:31    6
2016-03-27 12:50:22    5
2016-03-28 10:48:11    5
                      ..
2016-03-24 12:53:24    1
2016-03-16 11:47:03    1
2016-03-22 01:36:17    1
2016-03-29 10:57:13    1
2016-03-21 10:50:36    1
Name: date_crawled, Length: 280459, dtype: int64

In [903]:
autos["ad_created"].value_counts(normalize=True, dropna=False)

2016-04-03 00:00:00   0.04
2016-04-04 00:00:00   0.04
2016-03-20 00:00:00   0.04
2016-03-12 00:00:00   0.04
2016-03-21 00:00:00   0.04
                      ... 
2014-03-10 00:00:00   0.00
2016-01-03 00:00:00   0.00
2016-01-06 00:00:00   0.00
2015-12-06 00:00:00   0.00
2015-11-10 00:00:00   0.00
Name: ad_created, Length: 114, dtype: float64

In [904]:
autos["last_seen"].value_counts(dropna=False)

2016-04-06 13:45:54    17
2016-04-07 06:45:59    17
2016-04-07 09:46:25    16
2016-04-06 13:15:54    16
2016-04-07 07:44:31    16
                       ..
2016-03-28 15:56:58     1
2016-03-27 12:44:20     1
2016-03-24 10:43:09     1
2016-03-22 15:17:38     1
2016-03-31 04:16:43     1
Name: last_seen, Length: 182771, dtype: int64

In [905]:
autos["registration_year"].value_counts(dropna=False)

2000    24540
1999    22764
2005    22309
2006    20225
2001    20215
        ...  
1919        1
1915        1
7800        1
1400        1
8455        1
Name: registration_year, Length: 155, dtype: int64

In [906]:
autos["registration_month"].value_counts( dropna=False)

0     37639
3     36162
6     33163
4     30915
5     30627
7     28954
10    27334
11    25487
12    25377
9     25070
1     24554
8     23763
2     22398
Name: registration_month, dtype: int64

In [907]:
autos["registration_year"].describe()

count   371,443.00
mean      2,004.58
std          92.88
min       1,000.00
25%       1,999.00
50%       2,003.00
75%       2,008.00
max       9,999.00
Name: registration_year, dtype: float64

As we can see, there are years registered like 1000 and 9999 that obviously are false, so we will have to miss that information.
Let's explore cars before 1900 and after 2018:
<!-- Because it's unlikely that any car in this data set was registered before  -->

In [908]:
pd.to_numeric(autos["registration_year"])

autos_before_1900 = autos[autos["registration_year"] <= 1900]
autos_before_1900.describe()

price  registration_year  power_ps  kilometer  registration_month
count      68.00              68.00     68.00      68.00               68.00
mean   18,665.13           1,178.21    191.38  13,161.76                0.93
std   122,619.91             262.74    460.26  32,995.98                1.93
min         0.00           1,000.00      0.00   5,000.00                0.00
25%        15.25           1,000.00      0.00   5,000.00                0.00
50%       400.00           1,000.00      0.00   5,000.00                0.00
75%     1,058.50           1,266.25     18.50   5,000.00                0.25
max   999,999.00           1,800.00  1,800.00 150,000.00                7.00

In [909]:
autos_before_1900

date_crawled  \
649     2016-03-16 16:55:09   
16859   2016-03-29 23:42:16   
19125   2016-03-09 17:49:44   
35964   2016-04-01 06:02:10   
40733   2016-03-19 23:54:40   
...                     ...   
339167  2016-03-26 20:58:27   
341569  2016-03-30 11:36:35   
357294  2016-04-04 23:55:47   
365751  2016-03-22 00:38:15   
368731  2016-03-12 00:57:39   

                                                     name  seller offer_type  \
649                           195/55_r15_sommerreifen_alu  privat    Angebot   
16859                  Ford_mondeo_mk3_2.0tdc?_tubo_lader  privat    Angebot   
19125                              Subaru_Impreza_Wrx_2.0  privat    Angebot   
35964                                          UNFAL_Auto  privat    Angebot   
40733                         Suche_einen_Trabant_mit_12v  privat    Angebot   
...                                                   ...     ...        ...   
339167                       Audi_tt_8n_grill_ohne_emblem  privat    Angebot   
341569  MB_Alufelgen_mit_Continental_Sport_Contact_2__...  privat    Angebot   
357294                                    A3_zu_verkaufen  privat    Angebot   
365751                       Kostenlose_altautoentsorgung  privat    Angebot   
368731            Fiat_500c_Belvedere__Topolino__Oldtimer  privat    Angebot   

        price   abtest vehicle_type  registration_year  gearbox  power_ps  \
649         0  control          NaN               1111      NaN         0   
16859     190  control          NaN               1000      NaN         0   
19125    7999     test          NaN               1500  manuell       224   
35964     450  control          NaN               1800      NaN      1800   
40733    1234  control          NaN               1234      NaN        26   
...       ...      ...          ...                ...      ...       ...   
339167     30  control          NaN               1000      NaN         0   
341569    400     test          NaN               1000      NaN         0   
357294    700  control          NaN               1600  manuell      1600   
365751      1  control          NaN               1000      NaN      1000   
368731  11500     test          NaN               1800      NaN        16   

          model  kilometer  registration_month fuel_type           brand  \
649         NaN       5000                   0       NaN            opel   
16859    mondeo       5000                   0       NaN            ford   
19125   impreza       5000                   3       NaN          subaru   
35964       NaN       5000                   2       NaN      mitsubishi   
40733       NaN       5000                   1       NaN         trabant   
...         ...        ...                 ...       ...             ...   
339167      NaN       5000                   0       NaN            audi   
341569      NaN       5000                   0       NaN   mercedes_benz   
357294       a3     150000                   4    benzin            audi   
365751      NaN     150000                   0       NaN  sonstige_autos   
368731   andere       5000                   6    benzin            fiat   

       unrepaired_damage           ad_created postal_code            last_seen  
649                  NaN  2016-03-16 00:00:00       44628  2016-03-20 16:44:37  
16859                NaN  2016-03-29 00:00:00       47166  2016-04-06 10:44:58  
19125                NaN  2016-03-09 00:00:00       53577  2016-03-15 05:15:17  
35964               nein  2016-04-01 00:00:00       63322  2016-04-01 09:42:30  
40733                NaN  2016-03-19 00:00:00       38518  2016-04-06 03:15:22  
...                  ...                  ...         ...                  ...  
339167               NaN  2016-03-26 00:00:00       37574  2016-04-06 12:17:27  
341569               NaN  2016-03-30 00:00:00        8060  2016-04-01 06:16:46  
357294              nein  2016-04-04 00:00:00       86343  2016-04-05 06:44:07  
365751               NaN  2016-03

In [910]:
autos_after_2020 = autos[autos["registration_year"] >= 2020]
autos_after_2020.describe()

price  registration_year  power_ps  kilometer  registration_month
count     114.00             114.00    114.00     114.00              114.00
mean   14,957.96           6,502.53    110.43   9,956.14                1.46
std    94,347.02           2,699.80    547.77  15,985.28                3.07
min         0.00           2,066.00      0.00   5,000.00                0.00
25%        59.25           4,575.00      0.00   5,000.00                0.00
50%       799.50           6,000.00      0.00  10,000.00                0.00
75%     6,263.00           9,394.75     10.50  10,000.00                0.00
max   999,999.00           9,999.00  5,000.00 150,000.00               12.00

In [911]:
autos_after_2020

date_crawled  \
13576   2016-03-29 18:39:40   
15894   2016-03-14 00:52:02   
16657   2016-04-02 11:55:48   
18121   2016-03-23 16:43:29   
18196   2016-03-06 16:06:20   
...                     ...   
351235  2016-03-20 17:53:51   
352029  2016-03-09 07:01:27   
355280  2016-03-24 19:49:36   
357075  2016-04-02 17:44:03   
358386  2016-03-28 17:37:30   

                                                     name  seller offer_type  \
13576   Verkaufe_DESIGN_Streifen_/_Aufkleber_VW__Opel_...  privat    Angebot   
15894                                      Alles_anbieten  privat    Angebot   
16657                    quad_neu_t?v_tauschen_gegen_Auto  privat    Angebot   
18121                       Opel_GT_Karosserie_mit_Brief!  privat    Angebot   
18196               Reo_Vorkriegs_Oldtimer_Rennwagen_1928  privat    Angebot   
...                                                   ...     ...        ...   
351235                                Vw_Golf_1_16v_Turbo  privat    Angebot   
352029       Mini_MK2_Top_ohne_Rost_sucht_neuen_Liebhaber  privat    Angebot   
355280                      Golf_4_Karosse_schrott_Kupfer  privat    Angebot   
357075                             Tausche_hier_mein_Quad  privat    Angebot   
358386                                         Suche_Auto  privat    Angebot   

        price   abtest vehicle_type  registration_year  gearbox  power_ps  \
13576      49  control          NaN               5000      NaN         0   
15894       0  control          NaN               9999      NaN         0   
16657    1700     test          NaN               3200      NaN         0   
18121     700     test          NaN               9999      NaN         0   
18196    6500  control          NaN               8888      NaN         0   
...       ...      ...          ...                ...      ...       ...   
351235  12000     test          NaN               4000      NaN       500   
352029      0  control          NaN               7500  manuell         0   
355280     50  control          NaN               3000      NaN      3000   
357075      0  control          NaN               3500  manuell        75   
358386      1     test          NaN               3000      NaN         0   

         model  kilometer  registration_month fuel_type           brand  \
13576     golf       5000                  12       NaN      volkswagen   
15894      NaN      10000                   0       NaN  sonstige_autos   
16657      NaN       5000                   0       NaN  sonstige_autos   
18121   andere      10000                   0       NaN            opel   
18196      NaN      10000                   0       NaN  sonstige_autos   
...        ...        ...                 ...       ...             ...   
351235    golf       5000                   0       NaN      volkswagen   
352029  andere      10000                   0    benzin            mini   
355280    golf     100000                   6       NaN      volkswagen   
357075     NaN       5000                   3    benzin  sonstige_autos   
358386  zafira       5000                   0       NaN            opel   

       unrepaired_damage           ad_created postal_code            last_seen  
13576                NaN  2016-03-29 00:00:00       74523  2016-04-06 04:16:14  
15894                NaN  2016-03-13 00:00:00       32689  2016-03-21 23:46:46  
16657                NaN  2016-04-02 00:00:00       33649  2016-04-06 09:46:13  
18121                NaN  2016-03-23 00:00:00       21769  2016-04-05 20:16:15  
18196                NaN  2016-03-06 00:00:00       55262  2016-03-30 20:46:55  
...                  ...                  ...         ...                  ...  
351235              nein  2016-03-20 00:00:00       57392  2016-04-07 00:46:30  
352029              nein  2016-03-09 00:00:00        9669  2016-03-19 19:44:50  
355280                ja  2016-03-24 00:00:00       23992  2016-04-03 13:17:57  
357075               NaN  2016-04-02 00:00:00

We observe that cars before 1900 have a registration of 1800 as much (one century before the car was invented aproximmately) and cars after 2020 have a registration year from the future (obviously, it's impossible).
The problem is that most of them seem to be real data but only that column have a wrong data point. It's better to replace that info for null values.

In [912]:
bool_registration_year = (autos["registration_year"] < 1900) | (autos["registration_year"] > 2020)
autos.loc[bool_registration_year, "registration_year"] = np.nan

In [913]:
autos_year_nan = autos[autos["registration_year"].isnull()]
autos_year_nan

date_crawled  \
649     2016-03-16 16:55:09   
13576   2016-03-29 18:39:40   
15894   2016-03-14 00:52:02   
16657   2016-04-02 11:55:48   
16859   2016-03-29 23:42:16   
...                     ...   
357075  2016-04-02 17:44:03   
357294  2016-04-04 23:55:47   
358386  2016-03-28 17:37:30   
365751  2016-03-22 00:38:15   
368731  2016-03-12 00:57:39   

                                                     name  seller offer_type  \
649                           195/55_r15_sommerreifen_alu  privat    Angebot   
13576   Verkaufe_DESIGN_Streifen_/_Aufkleber_VW__Opel_...  privat    Angebot   
15894                                      Alles_anbieten  privat    Angebot   
16657                    quad_neu_t?v_tauschen_gegen_Auto  privat    Angebot   
16859                  Ford_mondeo_mk3_2.0tdc?_tubo_lader  privat    Angebot   
...                                                   ...     ...        ...   
357075                             Tausche_hier_mein_Quad  privat    Angebot   
357294                                    A3_zu_verkaufen  privat    Angebot   
358386                                         Suche_Auto  privat    Angebot   
365751                       Kostenlose_altautoentsorgung  privat    Angebot   
368731            Fiat_500c_Belvedere__Topolino__Oldtimer  privat    Angebot   

        price   abtest vehicle_type  registration_year  gearbox  power_ps  \
649         0  control          NaN                nan      NaN         0   
13576      49  control          NaN                nan      NaN         0   
15894       0  control          NaN                nan      NaN         0   
16657    1700     test          NaN                nan      NaN         0   
16859     190  control          NaN                nan      NaN         0   
...       ...      ...          ...                ...      ...       ...   
357075      0  control          NaN                nan  manuell        75   
357294    700  control          NaN                nan  manuell      1600   
358386      1     test          NaN                nan      NaN         0   
365751      1  control          NaN                nan      NaN      1000   
368731  11500     test          NaN                nan      NaN        16   

         model  kilometer  registration_month fuel_type           brand  \
649        NaN       5000                   0       NaN            opel   
13576     golf       5000                  12       NaN      volkswagen   
15894      NaN      10000                   0       NaN  sonstige_autos   
16657      NaN       5000                   0       NaN  sonstige_autos   
16859   mondeo       5000                   0       NaN            ford   
...        ...        ...                 ...       ...             ...   
357075     NaN       5000                   3    benzin  sonstige_autos   
357294      a3     150000                   4    benzin            audi   
358386  zafira       5000                   0       NaN            opel   
365751     NaN     150000                   0       NaN  sonstige_autos   
368731  andere       5000                   6    benzin            fiat   

       unrepaired_damage           ad_created postal_code            last_seen  
649                  NaN  2016-03-16 00:00:00       44628  2016-03-20 16:44:37  
13576                NaN  2016-03-29 00:00:00       74523  2016-04-06 04:16:14  
15894                NaN  2016-03-13 00:00:00       32689  2016-03-21 23:46:46  
16657                NaN  2016-04-02 00:00:00       33649  2016-04-06 09:46:13  
16859                NaN  2016-03-29 00:00:00       47166  2016-04-06 10:44:58  
...                  ...                  ...         ...                  ...  
357075               NaN  2016-04-02 00:00:00       96465  2016-04-04 15:17:51  
357294              nein  2016-04-04 00:00:00       86343  2016-04-05 06:44:07  
358386               NaN  2016-03-28 00:00:00       26624  2016-04-02 22:17:49  
365751               NaN  2016-03-21 00:00:00

In the case of registration_month we see that there are 37639 rows with the month 0. Obviously it doesn't exist therefore we do the same: convert this data point into null values.

In [914]:
bool_registration_month = autos["registration_month"] == 0
autos.loc[bool_registration_month, "registration_month"] = np.nan

In [915]:
autos_month_nan = autos[autos["registration_month"].isnull()]
autos_month_nan

date_crawled  \
0       2016-03-24 11:52:17   
9       2016-03-17 10:53:50   
15      2016-03-11 21:39:15   
16      2016-04-01 12:46:46   
36      2016-03-11 11:50:37   
...                     ...   
371460  2016-04-03 13:46:24   
371473  2016-03-15 19:57:11   
371482  2016-03-31 19:36:18   
371486  2016-03-30 20:55:30   
371494  2016-03-26 14:38:18   

                                                     name  seller offer_type  \
0                                              Golf_3_1.6  privat    Angebot   
9       VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...  privat    Angebot   
15                           KA_Lufthansa_Edition_450€_VB  privat    Angebot   
16                                            Polo_6n_1_4  privat    Angebot   
36                                       Opel_Kadett_E_CC  privat    Angebot   
...                                                   ...     ...        ...   
371460         Polo_g40_auch_Tausch_vag...no_vr6_gti_1.8t  privat    Angebot   
371473                                      Subaru_Allrad  privat    Angebot   
371482                                        Peugeot_206  privat    Angebot   
371486                                       Zu_verkaufen  privat    Angebot   
371494  BMW_760_erst_110_000_km_letze_Chance_bis_Monta...  privat    Angebot   

        price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0         480     test          NaN           1,993.00    manuell         0   
9         999     test   kleinwagen           1,998.00    manuell       101   
15        450     test   kleinwagen           1,910.00        NaN         0   
16        300     test          NaN           2,016.00        NaN        60   
36       1600  control       andere           1,991.00    manuell        75   
...       ...      ...          ...                ...        ...       ...   
371460   3500  control          NaN           1,995.00        NaN         0   
371473    400  control        kombi           1,991.00    manuell         0   
371482   1300  control   kleinwagen           1,999.00    manuell        75   
371486    350  control   kleinwagen           1,996.00        NaN        65   
371494  11500  control    limousine           2,004.00  automatik       445   

          model  kilometer  registration_month fuel_type       brand  \
0          golf     150000                 nan    benzin  volkswagen   
9          golf     150000                 nan       NaN  volkswagen   
15           ka       5000                 nan    benzin        ford   
16         polo     150000                 nan    benzin  volkswagen   
36       kadett      70000                 nan       NaN        opel   
...         ...        ...                 ...       ...         ...   
371460     polo     150000                 nan       NaN  volkswagen   
371473   legacy     150000                 nan    benzin      subaru   
371482  2_reihe     125000                 nan       NaN     peugeot   
371486    punto     150000                 nan       NaN        fiat   
371494      7er     125000                 nan    benzin         bmw   

       unrepaired_damage           ad_created postal_code            last_seen  
0                    NaN  2016-03-24 00:00:00       70435  2016-04-07 03:16:57  
9                    NaN  2016-03-17 00:00:00       27472  2016-03-31 17:17:06  
15                   NaN  2016-03-11 00:00:00       24148  2016-03-19 08:46:47  
16                   NaN  2016-04-01 00:00:00       38871  2016-04-01 12:46:46  
36                   NaN  2016-03-11 00:00:00        2943  2016-04-07 03:46:09  
...                  ...                  ...         ...                  ...  
371460               NaN  2016-04-03 00:00:00       74579  2016-04-05 12:44:38  
371473               NaN  2016-03-15 00:00:00       24558  2016-03-19 15:49:00  
371482               NaN  2016-03-31 00:00:00       35102  2016-04-06 13:44:44  
371486               NaN  2016-03-30 00:00:00       25436

In [916]:
# autos["registration_year"] = autos.registration_year.astype(str).str.split('.', expand = True)[0]

# autos["registration_month"] = autos.registration_month.astype(str).str.split('.', expand = True)[0]

In [917]:
autos["registration_year"].value_counts().sort_index(ascending=False)

2,019.00       26
2,018.00     3991
2,017.00    10542
2,016.00     9855
2,015.00     2982
            ...  
1,920.00        1
1,919.00        1
1,915.00        1
1,911.00        1
1,910.00      103
Name: registration_year, Length: 100, dtype: int64

In [918]:
autos["registration_month"].value_counts().sort_index(ascending=True)

1.00     24554
2.00     22398
3.00     36162
4.00     30915
5.00     30627
6.00     33163
7.00     28954
8.00     23763
9.00     25070
10.00    27334
11.00    25487
12.00    25377
Name: registration_month, dtype: int64

Next, we are going to analize the other date columns ('date_crawled', 'ad_created', 'last_seen').\
First, we create a function in order to convert the data in date format and then we create an empty dataframe to include the date columns in the format we want.

In [919]:
import datetime as dt

# creating function for conversion
date_format = '%Y-%m-%d'
def convert_dt(series):
    date = series.str[:10]
    date_dt = pd.to_datetime(date, format=date_format)
    return date_dt

# Creating an empty dataframe
autos_copy = pd.DataFrame()

# Adding columns to our new dataframe
autos_copy['date_crawled_dt'] = convert_dt(autos['date_crawled'])
autos_copy['ad_created_dt'] = convert_dt(autos['ad_created'])
autos_copy['last_seen_dt'] = convert_dt(autos['last_seen'])

Now, we analyze these series:

In [920]:
print(autos_copy['date_crawled_dt'].value_counts().sort_index(ascending=False))
print("\n")
print(autos_copy['date_crawled_dt'].describe(datetime_is_numeric=True))
print("\n")
print(autos_copy['ad_created_dt'].value_counts().sort_index(ascending=False))
print("\n")
print(autos_copy['ad_created_dt'].describe(datetime_is_numeric=True))
print("\n")
print(autos_copy['last_seen_dt'].value_counts().sort_index(ascending=False))
print("\n")
print(autos_copy['last_seen_dt'].describe(datetime_is_numeric=True))

2016-04-07      602
2016-04-06     1175
2016-04-05     4763
2016-04-04    13972
2016-04-03    14385
2016-04-02    13030
2016-04-01    12672
2016-03-31    11843
2016-03-30    12455
2016-03-29    12692
2016-03-28    13042
2016-03-27    11246
2016-03-26    11872
2016-03-25    12232
2016-03-24    11112
2016-03-23    11877
2016-03-22    12060
2016-03-21    13274
2016-03-20    13504
2016-03-19    13110
2016-03-18     4875
2016-03-17    11762
2016-03-16    11200
2016-03-15    12424
2016-03-14    13470
2016-03-13     5845
2016-03-12    13445
2016-03-11    12153
2016-03-10    12097
2016-03-09    12708
2016-03-08    12425
2016-03-07    13258
2016-03-06     5372
2016-03-05     9491
Name: date_crawled_dt, dtype: int64


count                           371443
mean     2016-03-20 21:49:56.290994944
min                2016-03-05 00:00:00
25%                2016-03-13 00:00:00
50%                2016-03-21 00:00:00
75%                2016-03-29 00:00:00
max                2016-04-07 00:00:00
Name: dat

We don't detect any information that seems to be wrong, we just learn that the crawler collected this data over a span of around a month in March/April of 2016.
In the three columns the last data coincides therefore we can assume that that is the last date we gathered information.
In this case, there are not outliers, so probably it's not necessary to clean the data. 

## Exploring Price by Brand

Firstly, we are going to explore the distribution of cars by brand: 

In [921]:
pd.options.display.float_format = '{:,.5f}'.format
distribution_autos_bybrand = autos["brand"].value_counts(normalize=True, dropna=False)*100
print(distribution_autos_bybrand)

volkswagen       21.43478
bmw              10.83989
opel             10.80435
mercedes_benz     9.50455
audi              8.84846
ford              6.88316
renault           4.83708
peugeot           2.96869
fiat              2.60444
seat              1.89047
mazda             1.53294
skoda             1.51867
smart             1.41314
citroen           1.39483
nissan            1.35606
toyota            1.26372
sonstige_autos    1.06773
hyundai           0.98158
mini              0.91373
volvo             0.89543
mitsubishi        0.82408
honda             0.76351
kia               0.68786
alfa_romeo        0.63105
suzuki            0.62648
porsche           0.59525
chevrolet         0.49644
chrysler          0.39091
dacia             0.24230
jeep              0.21726
daihatsu          0.21699
subaru            0.20972
land_rover        0.20730
jaguar            0.16719
trabant           0.15857
daewoo            0.14592
saab              0.14269
rover             0.13192
lancia      

For our study we are going to select only the brands that have at least 1% of presence in our dataset:

In [922]:
top_brands_1 = distribution_autos_bybrand.loc[distribution_autos_bybrand > 1]
print(top_brands_1)

top_brands = top_brands_1.index
print("\n")
print(top_brands)

volkswagen       21.43478
bmw              10.83989
opel             10.80435
mercedes_benz     9.50455
audi              8.84846
ford              6.88316
renault           4.83708
peugeot           2.96869
fiat              2.60444
seat              1.89047
mazda             1.53294
skoda             1.51867
smart             1.41314
citroen           1.39483
nissan            1.35606
toyota            1.26372
sonstige_autos    1.06773
Name: brand, dtype: float64


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'mazda', 'skoda', 'smart', 'citroen',
       'nissan', 'toyota', 'sonstige_autos'],
      dtype='object')


Now, we are going to create a dictionary to learn what is the average price of this selected brands:

In [923]:
pd.options.display.float_format = '{:,.2f}'.format
dict_brands = {}

for brand in top_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean = selected_rows["price"].mean()
    dict_brands[brand] = mean

In [924]:
dict_brands

{'volkswagen': 5183.8017784923,
 'bmw': 8268.30555831512,
 'opel': 2863.363350941892,
 'mercedes_benz': 8352.312287559484,
 'audi': 8768.335442845408,
 'ford': 3596.111667383737,
 'renault': 2334.5132186786886,
 'peugeot': 3166.9815906411536,
 'fiat': 2775.072048790573,
 'seat': 4356.66604955853,
 'mazda': 3925.1241657885494,
 'skoda': 6411.183478106719,
 'smart': 3531.5620118117736,
 'citroen': 3608.7967573827445,
 'nissan': 4527.5108199324995,
 'toyota': 5232.657008947593,
 'sonstige_autos': 13824.601109430156}

We can observe that after sonstige_autos (the most expensive brand)there are three brands that highlight above the rest of brands (BMW, Mercedes-Benz and Audi). Everybody knows that they are luxury brands of cars therefore our data makes sense.\
In spite of that, these three brands are among the most popular brands in our data set.
Opel, Ford and Smart are the cheapest brands of the data set.
Volkswagen, the most popular brand, is in between of them, near to the average price, probably.

Now, we are going to analyze the mileage of the top 6 brands to find out if there's any link with mean price:

In [925]:
top_6_brands = distribution_autos_bybrand.head(6)
print(top_6_brands)
top_six_brands = top_6_brands.index
print("\n")
print(top_six_brands)

volkswagen      21.43
bmw             10.84
opel            10.80
mercedes_benz    9.50
audi             8.85
ford             6.88
Name: brand, dtype: float64


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


For that, we are going to create a new dataframe with mean price and mileage of these brands:

In [926]:
# dict_avg_six_brands = {}

# for brand in top_six_brands:
#     selected_rows = autos[autos["brand"] == brand]
#     mean_price = selected_rows["price"].mean()
#     mean_kilometers = selected_rows["kilometer"].mean()
#     dict_avg_six_brands[brand] = mean_price, mean_kilometers

# print(dict_avg_six_brands)

In [927]:
dict_avg_price_six_brands = {}

for brand in top_six_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean_price = selected_rows["price"].mean()

    dict_avg_price_six_brands[brand] = mean_price

print(dict_avg_price_six_brands)

{'volkswagen': 5183.8017784923, 'bmw': 8268.30555831512, 'opel': 2863.363350941892, 'mercedes_benz': 8352.312287559484, 'audi': 8768.335442845408, 'ford': 3596.111667383737}


In [928]:
dict_avg__km_six_brands = {}

for brand in top_six_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean_kilometers = selected_rows["kilometer"].mean()
    dict_avg__km_six_brands[brand] = mean_kilometers

print(dict_avg__km_six_brands)

{'volkswagen': 128579.46695470874, 'bmw': 132768.47804490363, 'opel': 128907.72949267417, 'mercedes_benz': 130662.67278495355, 'audi': 129720.84461618036, 'ford': 123837.75961199985}


In [930]:
avg_p_series = pd.Series(dict_avg_price_six_brands)
avg_km_series = pd.Series(dict_avg__km_six_brands)
six_brands_df = pd.DataFrame(avg_p_series, columns=["price_mean"])
six_brands_df["kilometer_mean"] = avg_km_series
print(six_brands_df)

               price_mean  kilometer_mean
volkswagen       5,183.80      128,579.47
bmw              8,268.31      132,768.48
opel             2,863.36      128,907.73
mercedes_benz    8,352.31      130,662.67
audi             8,768.34      129,720.84
ford             3,596.11      123,837.76


In [933]:
print("The average of kilometers in the 6 main brands is ", round(six_brands_df["kilometer_mean"].mean(),2))

The average of kilometers in the 6 main brands is  129079.49


# Conclusions:

As we can see, the kilometer_mean in the 6 most popular brands is similar.
Although it's true that the kilometer_mean is higher in the brands with the higher price_mean, there is not significant correlation between price_mean and kilometer_mean.
Other conclusions could be:

The highest count frequency for kilometer is 150,000 km
The distribution of date_crawled is fairly consistent throughout the whole crawling period
The peak time for ads_created is March 2016 and beginning of April 2016
The distribution of last_seen percentage is rather consistent throughout the timeframe